sdk/src/types/block/core.rs

```diff
@@ -279,45 +480,130 @@ fn verify_parents(
+            Self::Validation(value.into())
+        }
+    }
+
+    impl<'de> Deserialize<'de> for BlockDataDto {
+        fn deserialize<D: serde::Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+            let value = Value::deserialize(d)?;
+            Ok(
+                match value
+                    .get("type")
+                    .and_then(Value::as_u64)
+                    .ok_or_else(|| serde::de::Error::custom("invalid block type"))? as u8
+                {
+                    BasicBlock::KIND => Self::Basic(
+                        BasicBlockDataDto::deserialize(value)
+                            .map_err(|e| serde::de::Error::custom(format!("cannot deserialize basic block: {e}")))?,
+                    ),
+                    ValidationBlock::KIND => {
+                        Self::Validation(ValidationBlockDataDto::deserialize(value).map_err(|e| {
+                            serde::de::Error::custom(format!("cannot deserialize validation block: {e}"))
+                        })?)
+                    }
+                    _ => return Err(serde::de::Error::custom("invalid block type")),
+                },
+            )
+        }
+    }
+
+    impl Serialize for BlockDataDto {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: serde::Serializer,
+        {
+            #[derive(Serialize)]
+            #[serde(untagged)]
+            enum BlockTypeDto_<'a> {
+                T0(&'a BasicBlockDataDto),
+                T1(&'a ValidationBlockDataDto),
+            }
+            #[derive(Serialize)]
+            struct TypedBlock<'a> {
+                #[serde(flatten)]
+                kind: BlockTypeDto_<'a>,
+            }
+            let output = match self {
+                Self::Basic(b) => TypedBlock {
+                    kind: BlockTypeDto_::T0(b),
+                },
+                Self::Validation(b) => TypedBlock {
+                    kind: BlockTypeDto_::T1(b),
+                },
+            };
+            output.serialize(serializer)
+        }
+    }
+
     /// The block object that nodes gossip around in the network.
     #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
     #[serde(rename_all = "camelCase")]
     pub struct BlockDto {
-        ///
         pub protocol_version: u8,
-        ///
-        pub strong_parents: BTreeSet<BlockId>,
-        pub weak_parents: BTreeSet<BlockId>,
-        pub shallow_like_parents: BTreeSet<BlockId>,
-        ///
-        #[serde(default, skip_serializing_if = "Option::is_none")]
-        pub payload: Option<PayloadDto>,
         #[serde(with = "string")]
-        pub burned_mana: u64,
+        pub network_id: u64,
+        #[serde(with = "string")]
+        pub issuing_time: u64,
+        pub slot_commitment_id: SlotCommitmentId,
+        pub latest_finalized_slot: SlotIndex,
+        pub issuer_id: IssuerId,
+        #[serde(flatten)]
+        pub data: BlockDataDto,
```

Comment:
It's not flattened and should be called `block`?

Reply:
fixed it